## Handwritten Digits Classification
### Using Convolutional Neural Network

We first create a basic convnet which is a stack of `Conv2D` and `MaxPooling2D` layers.

A CNN takes as input tensors of shape `(image_height, image_width, image_channels)`.

For our network we will configure the inputs of size `(28, 28, 1)`.


In [2]:
from keras import layers, models

model = models.Sequential()
model.add(layers.Conv2D(32, (3,3), activation = 'relu', input_shape = (28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation = 'relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation = 'relu'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 13, 13, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_4 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
Total params: 55744 (217.75 KB)
Trainable params: 5574

Now we need to feed the last ouput tensor into a densly-connected classifier network, like a stack of Dense layers.

These classifiers process vectors that are 1D, whereas our current output is 3D tensor.

Therefore, we will have to flatten our 3D outputs to 1D, and then add few Dense layers on top.

As we are going to do 10-way classification, we use a final layer with 10 outputs and a softmax activation.

In [3]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation = 'relu'))
model.add(layers.Dense(10, activation = 'softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPoolin  (None, 13, 13, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_4 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 576)              

### Use CNN on MNIST Digits

In [6]:
import keras
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32')/255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32')/255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)



In [9]:
model.compile(optimizer = 'rmsprop',
             loss = 'categorical_crossentropy',
             metrics = ['accuracy'])
model.fit(train_images, train_labels, epochs = 8, batch_size = 64)

Epoch 1/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0125 - accuracy: 0.9961
Epoch 2/8
938/938 [==============================] - 12s 12ms/step - loss: 0.0096 - accuracy: 0.9970
Epoch 3/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0077 - accuracy: 0.9976
Epoch 4/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0065 - accuracy: 0.9979
Epoch 5/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0049 - accuracy: 0.9988
Epoch 6/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0056 - accuracy: 0.9984
Epoch 7/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0037 - accuracy: 0.9990
Epoch 8/8
938/938 [==============================] - 12s 13ms/step - loss: 0.0037 - accuracy: 0.9987


We reached an accuracy of `0.9987( i.e, 99.87% )` on the training data. Now check the model's performance on testing data too.

In [10]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

test_acc

313/313 [==============================] - 1s 3ms/step - loss: 0.0408 - accuracy: 0.9922


0.9922000169754028

The test accuracy turns out to be `99.22%`, which is more than the accuracy of multilayer perceptron on test set.